pennylane/gradients/general_shift_rules.py

```diff
@@ -396,10 +397,11 @@ def _copy_and_shift_params(tape, indices, shifts, multipliers, cast=False):
 
         # Shift copied parameter
         new_params = list(op.data)
-        multiplier = qml.math.convert_like(multiplier, new_params[p_idx])
-        multiplier = qml.math.cast_like(multiplier, new_params[p_idx])
-        shift = qml.math.convert_like(shift, new_params[p_idx])
-        shift = qml.math.cast_like(shift, new_params[p_idx])
+        if not isinstance(new_params[p_idx], numbers.Integral):
+            multiplier = qml.math.convert_like(multiplier, new_params[p_idx])
+            multiplier = qml.math.cast_like(multiplier, new_params[p_idx])
+            shift = qml.math.convert_like(shift, new_params[p_idx])
+            shift = qml.math.cast_like(shift, new_params[p_idx])

```

Comment:
```suggestion
        if qml.math.get_interface(new_params[p_idx]) == "tensorflow":
            multiplier = qml.math.cast_like(multiplier, new_params[p_idx])
            shift = qml.math.cast_like(shift, new_params[p_idx])
        multiplier = qml.math.convert_like(multiplier, new_params[p_idx]) 
        shift = qml.math.convert_like(shift, new_params[p_idx])
```

Can't believe I didn't think of this before.

Reply:
Looks good, but will that work if `qml.math.get_interface(new_params[p_idx]) == "tensorflow"` and `new_params[p_idx]` is an integer? Or does that combination never happen?

Reply:
```suggestion
        if qml.math.get_interface(new_params[p_idx]) == "tensorflow":
            new_param_p_idx = qml.math.cast(new_params[p_idx], float)
       else:
           new_param_p_idx = new_params[p_idx]
        multiplier = qml.math.convert_like(multiplier, new_param_p_idx ) 
        shift = qml.math.convert_like(shift, new_param_p_idx )
```

Maybe?

Reply:
Just some context, but in the original implementation we tried to ensure we weren't causing TF warning due to incompatible casting, as we were getting lots of user bugs. cc @trbromley

Reply:
https://github.com/PennyLaneAI/pennylane/issues/3787

Reply:
To confirm @josh146's point, are we making sure that random TF warnings are not being encountered by PL users (with the TF interface)?

Reply:
The warning above is reproducible with
```
pytest -s tests/interfaces/default_qubit_2_integration/test_tensorflow_default_qubit_2.py -k 'TestTensorflowExecuteIntegration and test_ragged_differentiation'
```
 but I don't think it's a result of this PR. It shows in `6b149a19cd1c7100` and `v0.31.0`.